identity/app/controllers/EmailVerificationController.scala

```diff
@@ -24,28 +29,52 @@ class EmailVerificationController(
     with ImplicitControllerExecutionContext
     with SafeLogging {
 
-  import authenticatedActions.fullAuthWithIdapiUserAction
-
-  def completeRegistration(): Action[AnyContent] =
-    fullAuthWithIdapiUserAction.async { implicit request =>
-      val idRequest = idRequestParser(request)
+  def completeRegistration(): Action[AnyContent] = {
+    Action.async { implicit request =>
       val page = IdentityPage("/complete-registration", "Complete Signup", isFlow = true)
       val verifiedReturnUrlAsOpt = returnUrlVerifier.getVerifiedReturnUrl(request)
 
-      Future.successful(
-        Ok(
-          IdentityHtmlPage.html(
-            views.html.verificationEmailResent(
-              request.user,
-              idRequest,
-              idUrlBuilder,
-              verifiedReturnUrlAsOpt,
-              returnUrlVerifier.defaultReturnUrl,
-              isSignupFlow = true,
-            ),
-          )(page, request, context),
-        ),
-      )
+      def decryptEmail(token: Option[String]): Future[Option[String]] = {
+        token match {
+          case Some(token) => api.decryptEmailToken(token).map(_.toOption)
+          case _ => {
+            logger.error("encryptedEmail parameter was absent from complete-registration page")
+            Future.successful(None)
+          }
+        }
+      }
+
+      decryptEmail(request.getQueryString("encryptedEmail"))
+        .map(email =>
+          Ok(
+            IdentityHtmlPage.html(
+              views.html.verificationEmailResent(
+                idUrlBuilder,
+                verifiedReturnUrlAsOpt,
+                returnUrlVerifier.defaultReturnUrl,
+                email,
+              ),
+            )(page, request, context),
+          ),
+        )
     }
+  }
+
+  def resendValidationEmail(token: String): Action[AnyContent] = {
+    Action.async { implicit request =>
+      val response = for {
+        email <- EitherT(api.decryptEmailToken(token))
```

Comment:
it might be worth adding an idapi endpoint to send validation emails to an encrypted email address rather than making 3 API calls.